### ISSUE TITLE
I like several parts of this project. What's missing is a universal build system..

### ISSUE THREAD

**Issue by mkschreder** [giseburt/Motate#2]
I have been adopting OpenWRT/buildroot for this kind of bare metal development. It gives a versatile build system that is able to cross compile software for microcontrollers just as one would do when crosscompiling a linux distro. OpenWRT already handles dependencies, building from configure & make, cmake etc. I think it is a much better way to go because it allows true modularization of libraries and robust methods for crosscompiling third party code that one wants to link into ones firmware. Everything is built from source, including all the tools and even the toolchain (gcc). 

I'm putting it together here: https://github.com/mkschreder/openwrt-micro.git

Currently only working target is stm32f10x and flysky-fs-t6 board, but I will be adding more since it is just a matter of adding packages for various libs and firmwares. (I'm putting together everything for stm32 as a  reusable library here: https://github.com/mkschreder/libstm32.git).

Example: make ar-t6 firmware
make package/art6/compile (will build gcc, newlib etc only once and then will build the package provided that it has been selected in menuconfig).

**Comment by giseburt** [comment]
Hi @mkschreder,

Well, you've stumbled upon the old repo, and I realized I've not marked it as such, so I've done that now so other don't have that happen as well.

The new repo is at https://github.com/synthetos/Motate , and the current most active branch is the [sam4e-port](https://github.com/synthetos/Motate/blob/sam4e-port/). I hope to merge that to master in the coming weeks. 

That said, we try to have as few external dependencies in Motate as possible, since Motate is an external dependency itself. This also helps us manage how many other/external projects we have to track.

That said, we have made a build system that is as universal as possible, without having to build the world or have code generators or configuration scripts before the Makefiles. So, if you look at the parent project ([g2core](https://github.com/synthetos/g2/)) as an example, it has Motate as a git submodule and the [Makefile](https://github.com/synthetos/g2/blob/edge/g2core/Makefile) sets up the structure and then hands all of the rest of the build to Motate by including the [`Motate.mk`](https://github.com/synthetos/Motate/blob/sam4e-port/MotateProject/motate/Motate.mk) file.

I desperately need to document the Motate repo, but some of the build structure documentation can be found in the g2core wiki: [Project Structure](https://github.com/synthetos/g2/wiki/Project-Structure-and-Motate) and [Adding Boards](https://github.com/synthetos/g2/wiki/Adding-and-Revising-Boards)

Note that we do not build the tools, esp since we support OS X, Windows, and Linux, so we rely on there being pre-built tools in a downloadable form for the architecture in question. Currently we only support ARM. (We used to support AVR, but it was neglected long enough that I pulled most of it. If someone is willing to maintain that we'll put it back in. We still have the build system partitioned out so it can support non-ARM core architectures.) And ARM provides prebuild GGC bare-metal toolchains for our three primary host platforms, so we just download that locally and use that. This also allows us to guarantee that the compiler tools used are those that we tested with and reduces the unexpected moving parts. See the [Tools `Makefile`](https://github.com/synthetos/Motate/blob/sam4e-port/Tools/Makefile) for the code that grabs the toolchain.

Thank you for the interest, and I look forward to further discussion!
-Rob

PS: I'm going to close this issue, but if you wish to continue the discussion please open a new issue in https://github.com/synthetos/Motate/issues (and refer to this one so others can follow the thread.)
